sm-shop/src/main/java/com/salesmanager/shop/store/controller/order/facade/OrderFacadeImpl.java

```diff
@@ -1312,4 +1335,61 @@ public ReadableTransaction captureOrder(MerchantStore store, Order order, Custom
 
 	}
 
+	/**
+	 * This method decrements the inventory quantity based on the quantity in the shopping cart
+	 * Or, if after the quantity has been decremented, an error occurs in payment processing or for any other reason
+	 * the order processing doesn't succeed it will roll it back.
+	 * @param item
+	 * @param originalInventoryQuantities stores the product and the original values we need to keep track of to rollback if necessary
+	 * @return
+	 * @throws ServiceException
+	 */
+	private Boolean decrementInventory(
+			ShoppingCartItem item,
+			HashMap<Product, Integer> originalInventoryQuantities) throws ServiceException {
+
+		Product persistable = productService.getById(item.getProductId());
+		Set<ProductAvailability> availabilities = persistable.getAvailabilities();
+
+		Boolean hasBeenDecremented = false;
+		for(ProductAvailability availability : availabilities) {
+			if (availability.getRegion().equals(Constants.ALL_REGIONS)) {
+
+				if (availability.getProductQuantity() >= item.getQuantity()) {
+					originalInventoryQuantities.put(persistable, availability.getProductQuantity());
+
+					availability.setProductQuantity(availability.getProductQuantity() - item.getQuantity());
+
+					hasBeenDecremented = true;
+				} else {
+					throw new ServiceException("quantity to remove exceeded inventory");
+				}
+			}
+		}
+		if(hasBeenDecremented) {
+			productService.update(persistable);
+		}
+		return hasBeenDecremented;
+	}
+
+	/**
+	 * Rolls Back the changes to the product quantities if necessary
+	 * @param quantitiesForRollback
+	 * @throws ServiceException
+	 */
+	private void rollbackDecrementInventory(HashMap<Product, Integer> quantitiesForRollback) throws ServiceException {
+
+		for (Map.Entry<Product, Integer> entry : quantitiesForRollback.entrySet()) {
```

Comment:
I think you can use stream's from java 8 for this code. it will look much better.